### ISSUE TITLE
Standardize execCommand("insertParagraph") in inline editing host

### ISSUE THREAD

**Issue by masayuki-nakano** [w3c/editing#175]
Test case: https://jsfiddle.net/d_toybox/7w0m6722/11/

I think that `"insertParagraph"` command (including Enter key press) in an inline editing host such as `<span contenteditable>foo bar</span>` should cause inserting `<br>` element.

In the testcases, Chrome and Safari just inserted `<br>` element for Editor 1 and Editor 2. I guess, Editor 3 case is just their unexpected bug.

Edge doesn't work well in any cases. It inserts `<p>` or `<div>` into `<span>` element.

Firefox will insert `<br>` element in these cases at 60 or 61 (note that starting Firefox 60, default paragraph separator becomes `<div>`.  59 and earlier used `<br>` and we have a bug in 60. See https://bugzilla.mozilla.org/show_bug.cgi?id=1435123)

**Comment by johanneswilm** [comment]
@masayuki-nakano I agree with your suggestion, but I would prefer not to spend more time on execCommand, as we are really trying to move away from it, and we have seen that editors have stopped using it about a decade ago. It would be much more helpful if you for example could implement the beforeinput event.

**Comment by masayuki-nakano** [comment]
> but I would prefer not to spend more time on execCommand, as we are really trying to move away from it,

Is there another API to modify DOM tree like execCommand?

> and we have seen that editors have stopped using it about a decade ago.

Even if spec editors have already given up standardize with a spec, Mozilla cannot stop working on it since there are a lot of incompatibility especially with Blink/WebKit and that make Firefox users inconvenient on some web apps. So, when we change Firefox's behavior, we want to include new tests into wpt. But in my understanding, a browser vendor shouldn't do it without filing issues.  Therefore, I keep filing this kind of issues when I need to fix incompatibility issues of undocumented editor behavior.

> It would be much more helpful if you for example could implement the beforeinput event.

FYI: I'm working on `InputEvent.inputType` and after that, I'll work on implementing `beforeinput`. But unfortunately, `InputEvent.inputType` requires me to understand Gecko's editor really deeper and I'm working on refining Gecko's editor design for making possible to implement them. (And I'm still thinking we should implement which `beforeinput` event (Blink's vs. WebKit's) on Firefox since they are really different...)

**Comment by johanneswilm** [comment]
> Is there another API to modify DOM tree like execCommand?

I am not sure what you mean. What is being used by editors as of today is the standard API to change the DOM tree -- things like `insertAdjacentHTML`, `setAttribute`, `getAttribute`, `replaceChild`, etc. . Execcommand is still used in a few places where there is no alternative.

> Even if spec editors have already given up standardize with a spec, Mozilla cannot stop working on it since there are a lot of incompatibility especially with Blink/WebKit and that make Firefox users inconvenient on some web apps.

Ok, if it's important to Mozilla, I understand. Just be aware that JS editors have moved away from using it in most cases (save above-mentioned exceptions). Just making sure you are aware that it's not how JS editors manipulate the DOM in most cases.

> FYI: I'm working on InputEvent.inputType and after that, I'll work on implementing beforeinput. But unfortunately, InputEvent.inputType requires me to understand Gecko's editor really deeper and I'm working on refining Gecko's editor design for making possible to implement them. (And I'm still thinking we should implement which beforeinput event (Blink's vs. WebKit's) on Firefox since they are really different...)

Understood and thanks for that info. The level 2 of the implementation (what Webkit has) is what has been arrived at through the discussions about requirements on the part of JS editors. Level 1 is something Chrome decided to do because they could for various reasons not ship level 2. Level 1 can work as an additional aid for making an editor, but one really has to continue to work with the APIs already used now for most things. Webkit's implementation is both level 1 and level 2 compliant. I hope Chrome can eventually get to level 2 or we can agree on a new level 3 that everyone can agree on.

**Comment by masayuki-nakano** [comment]
> I am not sure what you mean. What is being used by editors as of today is the standard API to change the DOM tree -- things like insertAdjacentHTML, setAttribute, getAttribute, replaceChild, etc. . Execcommand is still used in a few places where there is no alternative.

Thanks, that's what I want to know.

> Understood and thanks for that info. The level 2 of the implementation (what Webkit has) is what has been arrived at through the discussions about requirements on the part of JS editors. Level 1 is something Chrome decided to do because they could for various reasons not ship level 2. Level 1 can work as an additional aid for making an editor, but one really has to continue to work with the APIs already used now for most things. Webkit's implementation is both level 1 and level 2 compliant. I hope Chrome can eventually get to level 2 or we can agree on a new level 3 that everyone can agree on.

Sigh... Although it's off topic, implementing Level 1 is looks reasonable for us since Safari's implementation looks odd in some cases. On the other hand, Level 1 cannot cancel inputs from keyboard, so, it cannot replace legacy keypress event...

**Comment by johanneswilm** [comment]
> Safari's implementation looks odd in some cases.

Ok, it would be interesting to hear about those cases. Maybe discussing those can lead us to a level 3 that is good for everyone. But let's make a new issue about that.
